File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
n = int(input())

graph = []

for _ in range(n):
    graph.append(list(map(int, input())))

dy = [1, 0, -1, 0]
dx = [0, 1, 0, -1]

visited = [[0 for _ in range(n)] for _ in range(n)]

result = []
def dfs(y, x):
    visited[y][x] = 1
    for i in range(len(dy)):
        ny = y + dy[i]
        nx = x + dx[i]
        if 0 <= ny < n and 0 <= nx < n:
            if graph[ny][nx] == 1 and visited[ny][nx] == 0:
                result[len(result) - 1] += 1
                dfs(ny, nx)

for i in range(n):
    for j in range(n):
          if graph[i][j] == 1 and visited[i][j] == 0:
              result.append(1)
              dfs(i, j)

result.sort()
print(len(result))
for i in result:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int N, cnt;
int yy, xx;
string x;

vector <string> graph;
vector <vector<bool>> visited;
vector <int> result;

int dy[4] = { 1, 0, -1, 0 };
int dx[4] = { 0, 1, 0, -1 };

void dfs(int y, int x) {
	visited[y][x] = true;

	for (int k = 0; k < 4; k++) {
		yy = y + dy[k];
		xx = x + dx[k];
		if (xx < 0 || xx >= N || yy < 0 || yy >= N || visited[yy][xx] || graph[yy][xx] == '0')
			continue;

		cnt += 1;
		dfs(yy, xx);
	}
}

int main(void) {
	cin.tie(0)->sync_with_stdio(false);

	cin >> N;

	for (int i = 0; i < N; i++) {
		cin >> x;
		graph.push_back(x);
	}

	visited = vector<vector<bool>>(N, vector<bool>(N, false));
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			if (!visited[i][j] && graph[i][j] == '1') {
				cnt = 1;
				dfs(i, j);
				result.push_back(cnt);
			}
		}
	}

	cout << result.size() << "\n";

	sort(result.begin(), result.end());

	for (int x : result)
		cout << x << "\n";

	return 0;
}